core/src/main/java/org/kiwix/kiwixmobile/core/webserver/ZimHostActivity.kt

```diff
@@ -0,0 +1,290 @@
+import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.adapter.BooksOnDiskListItem.BookOnDisk
+import java.util.ArrayList
+import javax.inject.Inject
+
+class ZimHostActivity : BaseActivity(), ZimHostCallbacks, ZimHostContract.View {
+  @Inject
+  internal lateinit var presenter: ZimHostContract.Presenter
+  @Inject
+  internal lateinit var alertDialogShower: AlertDialogShower
+  private lateinit var booksAdapter: BooksOnDiskAdapter
+  private lateinit var bookDelegate: BookOnDiskDelegate.BookDelegate
+  private var hotspotService: HotspotService? = null
+  private var ip: String? = null
+  private var serviceConnection: ServiceConnection? = null
+  private var progressDialog: ProgressDialog? = null
+  private val tag = "ZimHostActivity"
+  private val ipStateKey = "ip_state_key"
+
+  private val selectedBooksPath: ArrayList<String>
+    get() {
+      val selectedBooksPath = ArrayList<String>()
+      for (item in booksAdapter.items) {
+        if (item.isSelected) {
+          val (_, _, file) = item as BooksOnDiskListItem.BookOnDisk
+          selectedBooksPath.add(file.absolutePath)
+          Log.v(tag, "ZIM PATH : " + file.absolutePath)
+        }
+      }
+      return selectedBooksPath
+    }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_zim_host)
+
+    setUpToolbar()
+
+    bookDelegate = BookOnDiskDelegate.BookDelegate(sharedPreferenceUtil, null, null,
+      { bookOnDiskItem ->
+        select(bookOnDiskItem)
+        Unit
+      })
+    bookDelegate.selectionMode = SelectionMode.MULTI
+    booksAdapter = BooksOnDiskAdapter(
+      bookDelegate,
+      BookOnDiskDelegate.LanguageDelegate
+    )
+    if (savedInstanceState != null) {
+      ip = savedInstanceState.getString(ipStateKey)
+      layoutServerStarted()
+    }
+    recyclerViewZimHost.adapter = booksAdapter
+    presenter.attachView(this)
+
+    serviceConnection = object : ServiceConnection {
+
+      override fun onServiceConnected(className: ComponentName, service: IBinder) {
+        hotspotService = (service as HotspotService.HotspotBinder).service
+        hotspotService!!.registerCallBack(this@ZimHostActivity)
+      }
+
+      override fun onServiceDisconnected(arg0: ComponentName) {}
+    }
+
+    startServerButton.setOnClickListener { startStopServer() }
+  }
+
+  private fun startStopServer() {
+    when {
+      ServerUtils.isServerStarted -> stopServer()
+      selectedBooksPath.size > 0 -> startHotspotManuallyDialog()
+      else -> Toast.makeText(
+        this,
+        R.string.no_books_selected_toast_message,
+        Toast.LENGTH_SHORT
+      ).show()
+    }
+  }
+
+  private fun stopServer() {
+    startService(createHotspotIntent(ACTION_STOP_SERVER))
+  }
+
+  private fun select(bookOnDisk: BooksOnDiskListItem.BookOnDisk) {
+    val booksList: ArrayList<BooksOnDiskListItem> = ArrayList()
```

Comment:
we can go functional with `booksAdapter.items` instead of using a for loop

```
booksAdapter.items.map {
 if (item==it) it.copy(isSelected==!item.isSelected)
else it
}
```

Reply:
https://github.com/kiwix/kiwix-android/pull/1569/commits/8cb2c3a6057442128d22ac79c1b467f6ba23641d